File: app/static/examen.js
// El examen que se rinde en la plataforma.
//
// Tres responsabilidades: guardar lo escrito en el servidor mientras se trabaja, mostrar
// cuánto tiempo queda, y dejar constancia de dos cosas que pasan en la pantalla.
//
// Se bloquea pegar texto y se pide pantalla completa. Del resto no se puede: una página web
// no puede impedir que alguien cambie de aplicación, lea de otra pantalla o tenga el
// teléfono al lado, y fingir lo contrario sería peor que no hacer nada, porque daría una
// seguridad que no existe. Lo que sí se hace es que no pueda decir que no se enteró.
(function () {
  const cab = document.querySelector("[data-examen]");
  if (!cab) return;
  const aid = cab.dataset.examen;
  const base = cab.dataset.base || "";
  const form = document.getElementById("form-examen");
  const elGuardado = document.getElementById("guardado");
  const elReloj = document.getElementById("reloj");
  const elAviso = document.getElementById("aviso-incidente");
  const elConexion = document.getElementById("aviso-conexion");
  const campos = [...form.querySelectorAll("[data-orden]")];

  // ---------------------------------------------------------------- guardado
  //
  // Lo escrito vive en el servidor: si se corta la luz o se cierra la pestaña, quien
  // vuelve —acá o en otra máquina— encuentra lo que había. Por eso se guarda seguido y
  // no solo al entregar.
  let sucio = false, guardando = false, fallos = 0;

  function respuestas() {
    const r = {};
    campos.forEach(c => {
      if (c.type === "radio") { if (c.checked) r[c.dataset.orden] = c.value; }
      else if (c.value.trim()) r[c.dataset.orden] = c.value;
    });
    return r;
  }

  function hora() {
    return new Date().toLocaleTimeString("es-AR", { hour: "2-digit", minute: "2-digit" });
  }

  async function guardar() {
    if (!sucio || guardando) return;
    guardando = true;
    const enviado = respuestas();
    try {
      const r = await fetch(`${base}/examen/${aid}/guardar`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ respuestas: enviado }),
      });
      if (!r.ok) throw new Error(String(r.status));
      sucio = false; fallos = 0;
      elConexion.hidden = true;
      elGuardado.textContent = "Guardado " + hora();
      elGuardado.classList.remove("sin-guardar");
    } catch (e) {
      // Se reintenta solo: el próximo ciclo lo vuelve a tomar porque `sucio` sigue en pie.
      fallos++;
      if (fallos >= 2) elConexion.hidden = false;
      elGuardado.textContent = "Sin guardar";
      elGuardado.classList.add("sin-guardar");
    } finally {
      guardando = false;
    }
  }

  campos.forEach(c => {
    c.addEventListener("input", () => { sucio = true; crecer(c); });
    c.addEventListener("change", () => { sucio = true; });
  });
  setInterval(guardar, 5000);
  window.addEventListener("blur", guardar);

  // Los cuadros de respuesta crecen con lo que se escribe: una barra de scroll adentro de
  // un campo obliga a releer a ciegas lo que uno acaba de escribir.
  function crecer(c) {
    if (c.tagName !== "TEXTAREA") return;
    c.style.height = "auto";
    c.style.height = Math.max(c.scrollHeight + 2, 120) + "px";
  }
  campos.forEach(crecer);

  // ---------------------------------------------------------------- incidentes
  //
  // El aviso es inmediato a propósito: enterarse al final de que algo quedó anotado sería
  // una trampa; enterarse en el momento permite corregir la conducta o pedir explicaciones.
  // Y por eso la salida tapa el examen hasta que se la cierre, en vez de una banda que se
  // puede pasar por alto si uno está escribiendo más abajo.
  const elBloqueo = document.getElementById("bloqueo");
  const elBloqueoTitulo = document.getElementById("bloqueo-titulo");
  const elBloqueoTexto = document.getElementById("bloqueo-texto");
  const elBloqueoCerrar = document.getElementById("bloqueo-cerrar");
  const elContador = document.getElementById("contador");

  // Salir de la pantalla. Se pide que dure algo —dos segundos— para no anotar el
  // parpadeo de una notificación del sistema o un clic que devuelve el foco enseguida.
  let salioEn = null;
  const MINIMO_SALIDA = 2000;
  let cerrando = false;          // entregando o yéndose: lo que pase después ya no cuenta
  let pantallaPerdida = false;   // se salió de pantalla completa mientras estaba afuera

  function avisar(texto) {
    elAviso.textContent = texto;
    elAviso.hidden = false;
  }

  // ---------------------------------------------------------------- pantalla completa
  //
  // El examen se rinde en pantalla completa: ahí no hay barra de pestañas ni botones de
  // minimizar a la vista, y salir de ella es algo que el navegador avisa. No impide nada
  // —un segundo monitor o un teléfono siguen ahí—, pero el «miré otra ventana de reojo»
  // pasa a ser un acto deliberado que queda anotado.
  //
  // Solo donde el navegador la ofrece. En el iPhone no existe para páginas, y dejar a
  // alguien sin rendir por su teléfono sería castigar a quien menos tiene.
  const elPuerta = document.getElementById("puerta");
  const conPantalla = !!(document.fullscreenEnabled && document.documentElement.requestFullscreen);

  function enPantalla() { return !!document.fullscreenElement; }
  function pedirPantalla() {
    if (!conPantalla || enPantalla()) return Promise.resolve();
    return document.documentElement.requestFullscreen({ navigationUI: "hide" }).catch(() => {});
  }

  if (!conPantalla) {
    elPuerta.hidden = true;
  } else {
    document.getElementById("puerta-entrar").addEventListener("click", () => {
      pedirPantalla().then(() => { elPuerta.hidden = true; });
    });
    document.addEventListener("fullscreenchange", () => {
      if (enPantalla() || cerrando || !elPuerta.hidden) return;
      // Si además está fuera de la pestaña, la salida se anota por el otro camino cuando
      // vuelva; acá solo se recuerda que perdió la pantalla completa.
      if (salioEn !== null) { pantallaPerdida = true; return; }
      registrar("pantalla", {});
    });
  }

  // Volver de afuera tapa el examen hasta que se cierre con un clic. Sin temporizador: si
  // la salida fue un falso positivo, no le cuesta tiempo a quien no hizo nada. El botón
  // devuelve a pantalla completa cuando se la perdió: es un clic, y eso el navegador lo acepta.
  function bloquear(tipo, detalle) {
    elBloqueoTitulo.textContent = tipo === "pantalla"
      ? "Saliste de la pantalla completa" : "Saliste de la pantalla del examen";
    elBloqueoTexto.textContent = detalle || "";
    elBloqueoTexto.hidden = !detalle;
    elBloqueoCerrar.textContent = conPantalla && !enPantalla()
      ? "Volver al examen en pantalla completa" : "Volver al examen";
    elBloqueo.hidden = false;
    elBloqueoCerrar.focus();
  }
  elBloqueoCerrar.addEventListener("click", () => {
    pedirPantalla().then(() => { elBloqueo.hidden = true; pantallaPerdida = false; });
  });

  function pintarContador(llevados) {
    if (!llevados) return;
    const s = llevados.salida || 0, p = llevados.pegado || 0, f = llevados.pantalla || 0;
    if (!s && !p && !f) { elContador.hidden = true; return; }
    const partes = [];
    if (s) partes.push(`${s} salida${s === 1 ? "" : "s"} de la pantalla`);
    if (f) partes.push(`${f} salida${f === 1 ? "" : "s"} de pantalla completa`);
    if (p) partes.push(`${p} intento${p === 1 ? "" : "s"} de pegar`);
    elContador.textContent = "Registrado en este examen: " + partes.join(" · ");
    elContador.hidden = false;
  }
  pintarContador({ salida: +elContador.dataset.salidas, pegado: +elContador.dataset.pegados,
                   pantalla: +elContador.dataset.pantallas });

  // Se registra con fetch y no con sendBeacon porque hace falta la respuesta: el servidor
  // devuelve cuántos van, y ese número es el que se muestra. La salida se envía al VOLVER,
  // con la pestaña ya visible, así que no hay riesgo de que el envío se cancele.
  function registrar(tipo, detalle) {
    return fetch(`${base}/examen/${aid}/incidente`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ tipo, detalle }),
    }).then(r => r.json()).then(d => {
      if (!d || !d.ok) return;
      pintarContador(d.llevados);
      if (tipo === "pegado") avisar(d.aviso); else bloquear(tipo, d.detalle);
    }).catch(() => {
      // Sin conexión no se pierde el aviso, aunque el conteo quede para el próximo.
      if (tipo === "pegado") avisar("No se puede pegar texto en el examen. Quedó registrado el intento.");
      else bloquear(tipo, "");
    });
  }

  function seFue() { if (salioEn === null) salioEn = Date.now(); }
  function volvio() {
    if (salioEn === null) return;
    const fuera = Date.now() - salioEn;
    salioEn = null;
    if (fuera >= MINIMO_SALIDA) registrar("salida", { segundos: Math.round(fuera / 1000) });
    else if (pantallaPerdida) registrar("pantalla", {});   // fue corto, pero la pantalla completa se perdió
  }

  document.addEventListener("visibilitychange", () => {
    if (document.hidden) { seFue(); guardar(); } else { volvio(); }
  });
  window.addEventListener("blur", seFue);
  window.addEventListener("focus", volvio);
  form.addEventListener("submit", () => { cerrando = true; });
  window.addEventListener("pagehide", () => { cerrando = true; });

  // Pegar desde afuera: se impide y se anota el intento. Bloquear en el navegador no es
  // infranqueable —quien abra las herramientas de desarrollo lo saltea— pero convierte un
  // Cmd+V distraído en algo deliberado, que es exactamente lo que se busca. Se registra
  // cuánto, no qué: el texto es de la persona y no hay razón para guardarlo.
  //
  // El arrastrar-y-soltar va junto: es la otra forma de meter texto de afuera y no dispara
  // el mismo evento, así que bloquear solo el pegado dejaría la puerta de al lado abierta.
  campos.forEach(c => {
    c.addEventListener("paste", ev => {
      ev.preventDefault();
      const texto = (ev.clipboardData || window.clipboardData)?.getData("text") || "";
      registrar("pegado", { caracteres: texto.length, bloqueado: 1 });
    });
    c.addEventListener("drop", ev => {
      ev.preventDefault();
      const texto = ev.dataTransfer?.getData("text") || "";
      registrar("pegado", { caracteres: texto.length, bloqueado: 1 });
    });
  });

  // ---------------------------------------------------------------- reloj
  //
  // La cuenta va contra la hora del SERVIDOR: se calcula cuánto falta en el momento de
  // servir la página y desde ahí se descuenta. El reloj de la máquina de quien rinde
  // puede estar corrido, y con él se decidiría mal cuándo se acabó el examen.
  const cierre = cab.dataset.cierre;
  if (cierre) {
    const desfasaje = new Date(cab.dataset.ahora).getTime() - Date.now();
    const fin = new Date(cierre).getTime();
    let entregando = false;

    function tic() {
      const falta = fin - (Date.now() + desfasaje);
      if (falta <= 0) {
        elReloj.textContent = "0:00";
        if (!entregando) {
          entregando = true;
          // Se guarda antes de entregar para no perder lo último que se escribió, y
          // recién ahí se envía: lo que vale es lo que está en el servidor.
          sucio = true;
          cerrando = true;                   // salir de pantalla completa al entregar no es un incidente
          guardar().finally(() => {
            form.dataset.confirmado = "1";   // se acabó el tiempo: no hay nada que confirmar
            form.submit();
          });
        }
        return;
      }
      // Con más de un día por delante esto no es un examen con reloj, es una ventana: una
      // cuenta regresiva de tres cifras de horas no le dice nada a nadie. Va la fecha.
      if (falta > 24 * 3600 * 1000) {
        const d = new Date(fin);
        const dd = String(d.getDate()).padStart(2, "0"), mm = String(d.getMonth() + 1).padStart(2, "0");
        const hh = String(d.getHours()).padStart(2, "0"), mi = String(d.getMinutes()).padStart(2, "0");
        elReloj.textContent = `cierra el ${dd}/${mm} a las ${hh}:${mi}`;
        elReloj.classList.add("lejos");
        setTimeout(tic, 60000);
        return;
      }
      elReloj.classList.remove("lejos");
      const total = Math.floor(falta / 1000);
      const h = Math.floor(total / 3600), m = Math.floor((total % 3600) / 60), sg = total % 60;
      elReloj.textContent = h > 0
        ? `${h}:${String(m).padStart(2, "0")}:${String(sg).padStart(2, "0")}`
        : `${m}:${String(sg).padStart(2, "0")}`;
      elReloj.classList.toggle("apurado", total <= 300);
      setTimeout(tic, 1000);
    }
    tic();
  }

  // Guardar antes de irse, por si acaso. No se pregunta «¿seguro?»: el examen está en el
  // servidor, así que irse no pierde nada, y un cartel del navegador solo asustaría.
  window.addEventListener("pagehide", () => {
    if (!sucio) return;
    navigator.sendBeacon?.(`${base}/examen/${aid}/guardar`,
      new Blob([JSON.stringify({ respuestas: respuestas() })], { type: "application/json" }));
  });
})();

File: app/examen.py
"""El examen que se rinde acá adentro.

Hasta ahora una instancia se desarrollaba afuera —en Word, en papel, donde fuera— y el
sistema veía únicamente el archivo que llegaba al final. Cuando la instancia se marca
«en plataforma», el desarrollo pasa a ocurrir en una pantalla propia: se escribe o se
marca ahí, lo escrito se guarda en el servidor mientras se trabaja, y la entrega se arma
con eso.

Dos decisiones que conviene tener presentes al leer esto:

El borrador vive en el servidor y no en el navegador. Un examen de dos horas no puede
perderse porque se cortó la luz, se cerró una pestaña o se quedó sin batería: quien vuelve
—desde esa máquina o desde otra— encuentra lo que había escrito. Y lo que vale al cerrar
la ventana es lo que estaba guardado, no lo que quedó en una pantalla que nadie envió.

Los incidentes se registran, no bloquean. Una página web no puede impedir que alguien
cambie de aplicación, abra el teléfono o copie de otro lado; prometer lo contrario sería
vender una seguridad que no existe. Lo que sí se puede es dejar constancia de lo que
ocurre en la pantalla del examen, avisárselo a quien rinde en el momento, y mostrárselo
después a quien corrige. Es una señal para que la mire una persona, nunca un veredicto
automático.
"""
import json

from .db import get_db, utcnow

# Tipos de incidente que se registran. Deliberadamente pocos y concretos: cada uno tiene
# que poder explicarse en una frase a quien rinde y a quien corrige.
TIPOS = {
    "salida": "Salió de la pantalla del examen",
    "pegado": "Pegó texto desde otro lado",
}

# Cuánto puede seguir guardándose después de la hora de cierre. Es para la carrera entre
# el reloj y el último guardado —se escribe hasta el final y el envío tarda—, no para dar
# tiempo extra: el reloj del navegador y el del servidor nunca coinciden al segundo.
GRACIA_SEGUNDOS = 120


def borrador(db, user_id: int, assignment_id: int):
    return db.execute(
        "SELECT * FROM borradores WHERE user_id = ? AND assignment_id = ?",
        (user_id, assignment_id),
    ).fetchone()


def abrir(db, user_id: int, assignment_id: int):
    """El borrador de esta persona, creándolo si es la primera vez que entra.

    `iniciado_at` queda fijo desde acá: es cuándo empezó a rendir, y con eso la ficha
    puede decir cuánto tardó.
    """
    fila = borrador(db, user_id, assignment_id)
    if fila:
        return fila
    ahora = utcnow()
    db.execute(
        "INSERT INTO borradores (user_id, assignment_id, respuestas, iniciado_at, guardado_at)"
        " VALUES (?, ?, '{}', ?, ?)", (user_id, assignment_id, ahora, ahora),
    )
    return borrador(db, user_id, assignment_id)


def guardar(db, user_id: int, assignment_id: int, respuestas: dict):
    """Guarda lo escrito hasta ahora. No crea el borrador: si no existe, no hay examen abierto."""
    db.execute(
        "UPDATE borradores SET respuestas = ?, guardado_at = ?"
        " WHERE user_id = ? AND assignment_id = ?",
        (json.dumps(respuestas, ensure_ascii=False), utcnow(), user_id, assignment_id),
    )


def respuestas_de(fila) -> dict:
    """Las respuestas guardadas, con las claves como enteros. {} si no hay nada."""
    if not fila:
        return {}
    try:
        crudo = json.loads(fila["respuestas"] or "{}")
    except ValueError:
        return {}
    salida = {}
    for k, v in (crudo or {}).items():
        try:
            salida[int(k)] = v
        except (TypeError, ValueError):
            continue
    return salida


def borrar(db, user_id: int, assignment_id: int):
    db.execute("DELETE FROM borradores WHERE user_id = ? AND assignment_id = ?",
               (user_id, assignment_id))


# ------------------------------------------------------------------ incidentes

def registrar(db, user_id: int, assignment_id: int, tipo: str, detalle: dict) -> int:
    """Deja constancia de un incidente y devuelve cuántos van de ese tipo.

    El número vuelve a la pantalla para poder decirle a quien rinde «es la tercera vez»:
    avisar sin decir cuántas van invita a pensar que no se está contando.
    """
    if tipo not in TIPOS:
        return 0
    db.execute(
        "INSERT INTO incidentes (user_id, assignment_id, tipo, detalle, created_at)"
        " VALUES (?, ?, ?, ?, ?)",
        (user_id, assignment_id, tipo, json.dumps(detalle, ensure_ascii=False), utcnow()),
    )
    return db.execute(
        "SELECT COUNT(*) n FROM incidentes WHERE user_id = ? AND assignment_id = ? AND tipo = ?",
        (user_id, assignment_id, tipo),
    ).fetchone()["n"]


def colgar_de_entrega(db, user_id: int, assignment_id: int, submission_id: int):
    """Ata a la entrega los incidentes que todavía no eran de ninguna."""
    db.execute(
        "UPDATE incidentes SET submission_id = ?"
        " WHERE user_id = ? AND assignment_id = ? AND submission_id IS NULL",
        (submission_id, user_id, assignment_id),
    )


def de_entrega(db, submission_id: int):
    return db.execute(
        "SELECT * FROM incidentes WHERE submission_id = ? ORDER BY created_at",
        (submission_id,),
    ).fetchall()


def resumen(filas) -> dict:
    """Los incidentes contados y en una frase, para la ficha de la entrega."""
    salidas = [f for f in filas if f["tipo"] == "salida"]
    pegados = [f for f in filas if f["tipo"] == "pegado"]
    segundos = sum(_dato(f, "segundos") for f in salidas)
    caracteres = sum(_dato(f, "caracteres") for f in pegados)
    partes = []
    if salidas:
        partes.append(f"Salió de la pantalla {_veces(len(salidas))}"
                      + (f" ({_duracion(segundos)} en total)" if segundos else ""))
    if pegados:
        partes.append(f"Pegó texto {_veces(len(pegados))}"
                      + (f" ({caracteres} caracteres)" if caracteres else ""))
    return {"hay": bool(filas), "salidas": len(salidas), "pegados": len(pegados),
            "segundos": segundos, "caracteres": caracteres,
            "frase": " · ".join(partes) or "Sin incidentes registrados."}


def _dato(fila, clave: str) -> int:
    try:
        return int(json.loads(fila["detalle"] or "{}").get(clave, 0) or 0)
    except (ValueError, TypeError):
        return 0


def _veces(n: int) -> str:
    return "1 vez" if n == 1 else f"{n} veces"


def _duracion(segundos: int) -> str:
    if segundos < 60:
        return f"{segundos} s"
    minutos, resto = divmod(segundos, 60)
    return f"{minutos} min" + (f" {resto} s" if resto else "")


# ------------------------------------------------------- de lo guardado a la entrega

def texto_de(items, respuestas: dict) -> str:
    """Las respuestas escritas, numeradas como las espera la corrección del escrito.

    El formato es el mismo que si lo hubiera entregado en un archivo —«1. …», «2. …»—,
    así la puntuación por pregunta funciona igual sin saber por dónde entró la entrega.
    """
    partes = []
    for it in items:
        texto = (respuestas.get(it["orden"]) or "").strip()
        partes.append(f"{it['orden']}. " + (texto if texto else "(sin responder)"))
    return "\n\n".join(partes)


def marcadas_de(items, respuestas: dict) -> dict:
    """Las opciones elegidas, como las manda el formulario de marcado del panel."""
    salida = {}
    for it in items:
        letra = str(respuestas.get(it["orden"]) or "").strip().lower()[:1]
        if letra:
            salida[it["orden"]] = letra
    return salida


def hay_algo(items, respuestas: dict, tipo: str) -> bool:
    """¿Alcanza para entregar? Una hoja en blanco no se entrega sola al vencer el plazo."""
    if tipo == "choice":
        return bool(marcadas_de(items, respuestas))
    return any((respuestas.get(it["orden"]) or "").strip() for it in items)
